scipy/optimize/_zeros_py.py

```diff
@@ -1883,4 +1898,574 @@ def _scalar_optimization_prepare_result(work, res, res_work_pairs, active,
+          tolerance.
+        """
+        n = work.terms
+        n_new = n if work.nit == 0 else 1
+        il, ic, io = work.il, work.ic, work.io
+
+        # Central difference
+        # `work_fc` is *all* the points at which the function has been evaluated
+        # `fc` is the points we're using *this iteration* to produce the estimate
+        work_fc = (f[ic, :n_new], work.fs[ic, :], f[ic, -n_new:])
+        work_fc = np.concatenate(work_fc, axis=-1)
+        if work.nit == 0:
+            fc = work_fc
+        else:
+            fc = (work_fc[:, :n], work_fc[:, n:n+1], work_fc[:, -n:])
+            fc = np.concatenate(fc, axis=-1)
+
+        # One-sided difference
+        work_fo = np.concatenate((work.fs[io, :], f[io, :]), axis=-1)
+        if work.nit == 0:
+            fo = work_fo
+        else:
+            fo = np.concatenate((work_fo[:, 0:1], work_fo[:, -2*n:]), axis=-1)
+
+        work.fs = np.zeros((len(ic), work.fs.shape[-1] + 2*n_new))
+        work.fs[ic] = work_fc
+        work.fs[io] = work_fo
+
+        wc, wo = _differentiate_weights(work, n)
+        work.df_last = work.df.copy()
+        work.df[ic] = fc @ wc / work.h
+        work.df[io] = fo @ wo / work.h
+        work.df[il] *= -1
+
+        work.h /= work.fac
+        work.error_last = work.error
+        # Simple error estimate - the difference in derivative estimates between
+        # this iteration and the last. This is typically conservative because if
+        # convergence has begin, the true error is much closer to the difference
+        # between the current estimate and the *next* error estimate. However,
+        # we could use Richarson extrapolation to produce an error estimate that
+        # is one order higher, and take the difference between that and
+        # `work.df` (which would just be constant factor that depends on `fac`.)
+        work.error = abs(work.df - work.df_last)
+
+    def check_termination(work):
+        """Terminate due to convergence, non-finite values, or error increase"""
+        stop = np.zeros_like(work.df).astype(bool)
+
+        i = work.error < work.atol + work.rtol*abs(work.df)
+        work.status[i] = _ECONVERGED
+        stop[i] = True
+
+        if work.nit > 0:
+            i = ~((np.isfinite(work.x) & np.isfinite(work.df)) | stop)
+            work.df[i], work.status[i] = np.nan, _EVALUEERR
+            stop[i] = True
+
+        # With infinite precision, there is a step size below which
+        # all smaller step sizes will reduce the error. But in floating point
+        # arithmetic, catastrophic cancellation will begin to cause the error
+        # to increase again. This heuristic tries to avoid step sizes that are
+        # too small. Note that there are more theoretically sound approache,
+        # but this was simple and effective.
+        i = (work.error > work.error_last*10) & ~stop
+        work.status[i] = _EERRORINCREASE
+        stop[i] = True
```

Comment:
My note about "theoretically sound approaches" was referring to an old paper I saw about choosing the optimal step size to minimize total error - not the formula they give in textbooks that you can't apply unless you already know higher derivatives, but something algorithmic. I wasn't able to find that one again immediately, but I'm sure there are others out there.

I wouldn't expect the algorithm to detect convergence when the true derivative is zero with default settings. That's because the defaults require satisfaction of a relative error tolerance for termination, so anything other than the exact answer of zero would have infinite relative error.

If we allow an absolute tolerance:
```python3
from scipy.optimize._zeros_py import _differentiate

def f(x):
    return (x-1)**3

res = _differentiate(f, x=1, atol=1e-20)
res
#     success: True
#      status: 0
#          df: -1.4823076576950256e-21
#       error: 4.446922973085077e-21
#         nit: 7
#        nfev: 21
#           x: 1.0
```
it terminates successfully and gives an order-of-magnitude-accurate estimate of the absolute error. 

So I'll add a note that an absolute tolerance should be provided when the true error may be zero. As we discussed, let's consider other possible termination criteria and other ways to improve this (maybe measuring the actual step size is important?) in a followup.

Reply:
> With default settings, I wouldn't expect the algorithm to detect convergence when the true derivative is zero. That's because the defaults require satisfaction of a relative error tolerance for termination, so anything other than the exact answer of zero would have infinite relative error.

Right, I think it's just that with some tweaking we could allow even tighter `atol` values, since the pattern I've observed is the error increasing and then decreasing again. 

> I notice that this is not just restricted to central differences; forward and backward differences struggle with the same problem. Actually, they are bit worse off because they experience the problem even if the point is an extreme point, not a saddle point. Central differences get lucky for even-order polynomials like

Interesting, I didn't try that many examples, but found better convergence for the one-sided case for those I did try.

> Note that this doesn't seem to be an issue with _small magnitude derivatives_ in general -- mostly it's a problem with being close to a point at which the derivative is _exactly_ zero. For instance, when the function is scaled by a small magnitude but the requested derivative is not actually zero, it still satisfies the relative tolerance.

What's interesting to me is that for the cases I've tried, just removing the condition to stop if the error increases by more than an order of magnitude results in better convergence (at the expense of disaster for many other cases). It's not just a matter of not getting relative convergence, there are `atol` values which won't work because of the early stopping, but otherwise could. I'd like to wrap my head around this better. It's not a blocker for this PR though.